core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java

```diff
@@ -1050,7 +1058,12 @@ public static Expression convert(Expression operand, Type fromType,
           return Expressions.box(una.expression, toBox);
         }
       }
-      return Expressions.box(operand, toBox);
+      if (fromType == toBox.primitiveClass) {
+        return Expressions.box(operand, toBox);
+      }
+      return Expressions.box(
```

Comment:
Can we give more explain doc here, i understand this code id for the case `Integer.valueOf((int) longVal)`. But i still need to go through the code for a while why this code snippet is needed.

Reply:
Sure, I will add doc for explanation and example here.